### PR TITLE
Adding onTimeout function in Adapter Spec

### LINKED ISSUE #2254: Allow bidders to subscribe to own timed out bid data
## Type of issue
- Feature request

Add a function to the bidder spec definition that is called when a `BID_TIMEOUT` event occurs. Similar to `pbjs.onEvent('bidTimeout', (data) => ... )`, but the payload contains only the timed out bids for the subscribed bidder rather than all timed out bidders.

### PR DESCRIPTION
**Pull request by vedantseta**
## Type of change

- [ ] Feature
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?

## Description of change

This adds a function `onTimeout` to bidder spec, letting bidders to subscribe to their own timeout.

 Payload received to this function is same as `bidTimeout` event with timeout value and user configured params for this adunit and bidder combination.

 However the payload contains timed out bids for subscribed bidder, unlike `bidTimeout` event data.

## Example
```javascript
export const spec = {
        code: BIDDER_CODE,
        aliases: ['ex'], // short code
        isBidRequestValid: function(bid) {
            return !!(bid.params.placementId || (bid.params.member && bid.params.invCode));
        },
        onTimeout: function(data) {
             // BIDDER SPEFIC CODE HERE

        },
}

```

## Other information
This resolves #2254 
Documentation PR :  [prebid.github.io/pull/678](https://github.com/prebid/prebid.github.io/pull/678)

### CODE REVIEW

**Comment by vedantseta** [comment]
@matthewlane , Have changed `for..of` to `forEach` and have added documentation [prebid.github.io#678](https://github.com/prebid/prebid.github.io/pull/678) here. Can you please review the same?

File: src/utils.js [hunk 1]
```diff
@@ -846,7 +846,21 @@ export function getBidderRequest(bidRequests, bidder, adUnitCode) {
       .filter(bid => bid.bidder === bidder && bid.adUnitCode === adUnitCode).length > 0;
   }) || { start: null, auctionId: null };
 }
-
+/**
+ * Returns user configured bidder params from adunit
+ * @param {object} adunits
+ * @param {string} adunit code
+ * @param {string} bidder code
+ * @return {Array} user configured param for the given bidder adunit configuration
```
Comment (mkendall07) [inline src/utils.js]: it's really hard to visulize what this object looks like. Can you add it to the jsdocs typedef here?

File: src/adaptermanager.js [hunk 2]
```diff
@@ -437,3 +437,26 @@ exports.getBidAdapter = function(bidder) {
...
+  });
+  timedOutBidders = utils.groupBy(timedOutBidders, 'bidder');
+
+  Object.keys(timedOutBidders).forEach((bidder) => {
+    try {
+      const adapter = _bidderRegistry[bidder];
+      const spec = adapter.getSpec();
+      if (spec && spec.onTimeout && typeof spec.onTimeout === 'function') {
+        utils.logInfo(`Invoking ${bidder}.onTimeout`);
+        spec.onTimeout(timedOutBidders[bidder]);
```
Comment (mkendall07) [inline src/adaptermanager.js]: typedef would be nice here for the passed object.

File: src/utils.js (hunk 1 above)
Comment (vedantseta) [reply src/utils.js]: return type is Array of Objects, 
```javascript
const adUnits = [{
      code: 'adUnit1',
      bids: [{
        bidder: 'bidder1',
        params: {
          key1: 'value1'
        }
      }]
    }];
```
In this case object is adUnits[0].params, which is defined by user

**Pull request merged**